### PR TITLE
Stripping Empty Write Requests

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
To make the conformance tests pass, we should only send the absolute minimum request to the backend. For requests with transforms, if we can, we therefore should not send an empty write followed by the transforms.

This includes all changes from unmerged https://github.com/GoogleCloudPlatform/google-cloud-java/pull/2719. The relevant changes from this PR are in https://github.com/GoogleCloudPlatform/google-cloud-java/pull/2739/commits/ab04dbfa0dec3e4c34b95104562aa7910ca81737

This is based on https://github.com/googleapis/nodejs-firestore/pull/74

### CODE REVIEW

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 1]
```diff
@@ -0,0 +1,11292 @@
...
+      }
+    };
+
+    public TestCase
+    getTestCase() {
+      return TestCase.forNumber(
+          testCase_);
+    }
+
+    public static final int DESCRIPTION_FIELD_NUMBER = 1;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 2]
```diff
@@ -0,0 +1,11292 @@
...
+        case 4:
+          hash = (37 * hash) + SET_FIELD_NUMBER;
+          hash = (53 * hash) + getSet().hashCode();
+          break;
+        case 5:
+          hash = (37 * hash) + UPDATE_FIELD_NUMBER;
+          hash = (53 * hash) + getUpdate().hashCode();
+          break;
+        case 6:
+          hash = (37 * hash) + UPDATE_PATHS_FIELD_NUMBER;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 3]
```diff
@@ -0,0 +1,11292 @@
...
+      hash = (37 * hash) + DESCRIPTION_FIELD_NUMBER;
+      hash = (53 * hash) + getDescription().hashCode();
+      switch (testCase_) {
+        case 2:
+          hash = (37 * hash) + GET_FIELD_NUMBER;
+          hash = (53 * hash) + getGet().hashCode();
+          break;
+        case 3:
+          hash = (37 * hash) + CREATE_FIELD_NUMBER;
+          hash = (53 * hash) + getCreate().hashCode();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 4]
```diff
@@ -0,0 +1,11292 @@
...
+      if (testCase_ == 6) {
+        output.writeMessage(6, (com.google.cloud.firestore.conformance.TestDefinition.UpdatePathsTest) test_);
+      }
+      if (testCase_ == 7) {
+        output.writeMessage(7, (com.google.cloud.firestore.conformance.TestDefinition.DeleteTest) test_);
+      }
+      unknownFields.writeTo(output);
+    }
+
+    public int getSerializedSize() {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+          FirestoreOptions.newBuilder().setProjectId("projectID").build(),
+          Mockito.mock(FirestoreRpc.class));
+
+  @Captor private ArgumentCaptor<CommitRequest> commitCapture;
+
+  @Captor private ArgumentCaptor<BatchGetDocumentsRequest> getAllCapture;
+
+  @Captor private ArgumentCaptor<ApiStreamObserver> streamObserverCapture;
+
+  private Gson gson = new Gson();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+      byte[] buffer = new byte[length];
+      int read = inputStream.read(buffer);
+      Preconditions.checkState(
+          length == read,
+          "Encountered unexpected length for test case. Expected %d, but was %d",
+          length,
+          read);
+      TestDefinition.Test testDefinition = TestDefinition.Test.parseFrom(buffer);
+
+      if ((!includedTests.isEmpty() && !includedTests.contains(testDefinition.getDescription()))
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 7]
```diff
@@ -0,0 +1,11292 @@
...
+          hash = (37 * hash) + SET_FIELD_NUMBER;
+          hash = (53 * hash) + getSet().hashCode();
+          break;
+        case 5:
+          hash = (37 * hash) + UPDATE_FIELD_NUMBER;
+          hash = (53 * hash) + getUpdate().hashCode();
+          break;
+        case 6:
+          hash = (37 * hash) + UPDATE_PATHS_FIELD_NUMBER;
+          hash = (53 * hash) + getUpdatePaths().hashCode();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 8]
```diff
@@ -0,0 +1,11292 @@
...
+          break;
+        case 7:
+          hash = (37 * hash) + DELETE_FIELD_NUMBER;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import com.google.protobuf.NullValue;
+import com.google.protobuf.Timestamp;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/** Converts user input into the Firestore Value representation. */
+class UserDataConverter {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java]: Fixed.

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentMask.java [hunk 10]
```diff
@@ -41,9 +48,13 @@ static DocumentMask fromObject(Map<String, Object> values) {
     for (Map.Entry<String, Object> entry : values.entrySet()) {
       Object value = entry.getValue();
       FieldPath childPath = path.append(FieldPath.of(entry.getKey()));
-      if (value instanceof Map) {
+      if (entry.getValue() == FieldValue.SERVER_TIMESTAMP_SENTINEL) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/DocumentMask.java]: This is intentional.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 11]
```diff
@@ -0,0 +1,11292 @@
...
+          hash = (53 * hash) + getDelete().hashCode();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 12]
```diff
@@ -0,0 +1,11292 @@
...
+     * <code>.tests.DeleteTest delete = 7;</code>
+     */
+    public com.google.cloud.firestore.conformance.TestDefinition.DeleteTestOrBuilder getDeleteOrBuilder() {
+      if (testCase_ == 7) {
+         return (com.google.cloud.firestore.conformance.TestDefinition.DeleteTest) test_;
+      }
+      return com.google.cloud.firestore.conformance.TestDefinition.DeleteTest.getDefaultInstance();
+    }
+
+    private byte memoizedIsInitialized = -1;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 13]
```diff
@@ -0,0 +1,11292 @@
...
+     * <code>.tests.UpdateTest update = 5;</code>
+     */
+    public com.google.cloud.firestore.conformance.TestDefinition.UpdateTestOrBuilder getUpdateOrBuilder() {
+      if (testCase_ == 5) {
+         return (com.google.cloud.firestore.conformance.TestDefinition.UpdateTest) test_;
+      }
+      return com.google.cloud.firestore.conformance.TestDefinition.UpdateTest.getDefaultInstance();
+    }
+
+    public static final int UPDATE_PATHS_FIELD_NUMBER = 6;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 14]
```diff
@@ -0,0 +1,11292 @@
...
+      return result;
+    }
+
+    @java.lang.Override
+    public int hashCode() {
+      if (memoizedHashCode != 0) {
+        return memoizedHashCode;
+      }
+      int hash = 41;
+      hash = (19 * hash) + getDescriptor().hashCode();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+
+@RunWith(AllTests.class)
+public class ConformanceTest {
+  /** Interface implemented by the conformance test cases. */
+  private interface ConformanceTestCase extends Test, Describable {}
+
+  private static final String TEST_FILE = "./src/test/resources/test_data.binprotos";
+
+  /** Excluded tests by test description. */
+  private Set<String> excludedTests = Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+   * constants.
+   */
+  private Object convertValue(Object data) {
+    if (data instanceof Map) {
+      return convertMap((Map<String, Object>) data);
+    } else if (data instanceof List) {
+      return convertArray((List<Object>) data);
+    } else if (data.equals("Delete")) {
+      return FieldValue.delete();
+    } else if (data.equals("ServerTimestamp")) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+  /** Interface implemented by the conformance test cases. */
+  private interface ConformanceTestCase extends Test, Describable {}
+
+  private static final String TEST_FILE = "./src/test/resources/test_data.binprotos";
+
+  /** Excluded tests by test description. */
+  private Set<String> excludedTests = Collections.emptySet();
+
+  /** If non-empty, only runs tests included in this set. */
+  private Set<String> includedTests = Collections.emptySet();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 18]
```diff
@@ -0,0 +1,11292 @@
...
+     * <code>.tests.UpdatePathsTest update_paths = 6;</code>
+     */
+    public com.google.cloud.firestore.conformance.TestDefinition.UpdatePathsTestOrBuilder getUpdatePathsOrBuilder() {
+      if (testCase_ == 6) {
+         return (com.google.cloud.firestore.conformance.TestDefinition.UpdatePathsTest) test_;
+      }
+      return com.google.cloud.firestore.conformance.TestDefinition.UpdatePathsTest.getDefaultInstance();
+    }
+
+    public static final int DELETE_FIELD_NUMBER = 7;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 19]
```diff
@@ -0,0 +1,11292 @@
...
+    public com.google.cloud.firestore.conformance.TestDefinition.Test.TestCase getTestCase();
+  }
+  /**
+   * <pre>
+   * A Test describes a single client method call and its expected result.
+   * </pre>
+   *
+   * Protobuf type {@code tests.Test}
+   */
+  public  static final class Test extends
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+
+  @Spy
+  private FirestoreImpl firestoreMock =
+      new FirestoreImpl(
+          FirestoreOptions.newBuilder().setProjectId("projectID").build(),
+          Mockito.mock(FirestoreRpc.class));
+
+  @Captor private ArgumentCaptor<CommitRequest> commitCapture;
+
+  @Captor private ArgumentCaptor<BatchGetDocumentsRequest> getAllCapture;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java [hunk 21]
```diff
@@ -0,0 +1,11292 @@
...
+        return memoizedHashCode;
+      }
+      int hash = 41;
+      hash = (19 * hash) + getDescriptor().hashCode();
+      hash = (37 * hash) + DESCRIPTION_FIELD_NUMBER;
+      hash = (53 * hash) + getDescription().hashCode();
+      switch (testCase_) {
+        case 2:
+          hash = (37 * hash) + GET_FIELD_NUMBER;
+          hash = (53 * hash) + getGet().hashCode();
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/conformance/TestDefinition.java]: This is generated code.

File: google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/test/java/com/google/cloud/firestore/ConformanceTest.java]: Fixed.

**Comment by pongad** [comment]
LGTM. @garrettjonesgoogle @jba PTAL.

**Comment by jba** [comment]
I don't have anything to add.

**Comment by schmidt-sebastian** [comment]
Merged the other PR into this one. This should now be a lot more reviewable.

**Comment by garrettjonesgoogle** [comment]
There doesn't appear to be any (or much) impact on public surface - @pongad 's LGTM is sufficient.

**Pull request merged**